docs/FoLiA-idf.1
.TH FoLiA-idf 1 "2020 jan 19"

.SH NAME
FoLiA-idf - Folia IDF word counter
.SH SYNOPSIS
idf [options] DIR

.SH DESCRIPTION
.B FoLiA-idf
counts the intradocument frequency of the words in FoLiA documents.
It will process all folia XML files in DIR and store its results in the current directory in a file called DIR.idf.tsv

.SH OPTIONS
.B --clip
number
.RS
clipping factor. When an idf is lower than 'number', it will not be stored.
.RE

.B --lower
.RS
Lowercase all words.
.RE

.B --strings
.RS
search for <str> nodes in the FoLiA. The default is to search for <w> nodes.
.RE

.B -t
or
.B --threads
number
.RS
Number of concurrent threads the program may use
.RE

.B -V
or
.B --version
.RS
Show VERSION
.RE

.B -h
or
.B --help
.RS
get help
.RE

.B -e
expr
.RS
when searching for files,
.B
FoLiA-idf
will only consider files that match with the expression 'expr' which may contain wildcards. The 'expr' is only matched against the file part. Not against paths.
.RE

.B -R
.RS
when a DIR is provided,
.B FoLiA-idf
will recurse through this DIR and its subdirs to find files.
.RE

.SH BUGS
possibly

.SH AUTHORS
Ko van der Sloot

Martin Reynaert

email: lamasoftware@science.ru.nl

.SH SEE ALSO

.B FoLiA-stats


NEWS
foliautils 0.16  07-01-2021
[Ko vd Sloot]
* requires libfolia 2.7 or above
* provenance data is better for a lot of modules
* added better checking on invalid NCnames in some modules.
* FoLiA-abby:
  - a lot of refactoring and additions to handle font/style information
* FoLiA-pm:
  - Notes are handled correctly now
  - fixed error in xlink attributes
* FoLiA-page:
  - more types of Page files are handled now
  - fixed annotation declarations
  - fixed offset calculation (due to change in FoLiA's opinion on those)
  - page number is added as a <br> node and in the metadata
  - added a --trusttokens option. This means that Word items in the Page file
    are added as Word's in the FoLiA, embedded in Sentences.
  - added a --norefs option to avoid adding references to the original texts
* FoLiA-correct:
  - make sure that the default is to run on 1 thread
  - added a --rebase-inputclass option

* FoLiA-alto:
  - the -t option was not always handled correctly

[Maarten van Gompel]
* FoLiA-benchmark: guard against compiler optimisation #48

foliautils 0.15  15-09-2020
[Maarten van Gompel]
* FoLiA-txt: check if a string is empty after normalisation (fix for #46)
[Ko vd Sloot]
* folia-correct: fix one-off error in hemp handling (when no hemp was found) #45
* some refactoring
  - centralized definition of XML_PARSER_OPTIONS
* bugfix in threading

foliautils 0.14  15-04-2020
[Martin Reynaert]
* updated man pages

[Ko vd Sloot]
* added man pages
* revised usage() in many modules
* the default seperator in FoLiA-stats is '_' now
* fix for: https://github.com/LanguageMachines/foliautils/issues/37
* fix for: https://github.com/LanguageMachines/foliautils/issues/41
* adapted to changes in libfolia
* many small code refactorings
* FoLiA-correct is improved a lot, allowing ngram corrections in FoLiA
* FoLiA-stats accepts a 'word_in_doc' mode now
* FoLiA-alto by default ceated <w> nodes now. use --oldstring to get <str>
* improved a lot in tests/
* many small fixes

foliautils 0.13  30-08-2019
[Ko vd Sloot]
bug fixes:
* fix for https://github.com/LanguageMachines/foliautils/issues/35
* fix for https://github.com/LanguageMachines/foliautils/issues/36

[Maarten van Gompel]
new features:
* FoLiA-wordtranslate.cxx: frog should be able to deal with
  spaces now, no need for ugly non-breaking space hack that now causes
  other problems in frog (LanguageMachines/frog#77)
* FoLiA-wordtranslate.cxx: adding ability to constrain by language


foliautils 0.12  29-05-2019
[Ko vd Sloot]
Release for FoLiA 2.0.

foliautils 0.11  15-05-2019
[Ko vd Sloot]
Stabilizing release for pre FoLiA 2.0

* Updated and added some tests
* started moving common code to a separate file and build a library
  (libfoliautils)
  - hemp detection is one of them

* FoLiA-stats:
  - added possibility to read a list of directories + file-names to process
    into seperate output directories. (could be generalized to other programs)
  - better hemp detection
* FoLiA-correct:
  - use same hempo detections as FoLiA-stats
* FoLiA-abby:
  - support more flavors
* FoLiA-clean:
  - avoid removing the last remaining tekt on <w> nodes
  - cleaning of tokenization now works


foliautils 0.10  29-11-2018
[Ko vd Sloot]
  * fixed icu:namespace issues
  * added FoLiA-abby, an ABBY to FoLiA convertor
  * src/FoLiA-abby.cxx, src/FoLiA-page.cxx, src/FoLiA-pm.cxx:
     - Allow 'none' value for --prefix
  * src/FoLiA-page.cxx, src/FoLiA-hocr.cxx: fixed Alignment info
  * src/FoLiA-correct.cxx:
     - fixed a problem with correction of the last word of a trigram.
     - fix correction of paragraphs with only deeper text
     - The --rank option accepts more flavors of files
  * src/FoLiA-stats.cxx:
     - added a --detokenize option
  * several minor fixes, refactorings etc.
  * updated tests

foliautils 0.9.2  05-06-2018
[Ko vd Sloot]
Bug fix release:
 * append small prefixes to output filenames, to ALWAYS avoid names starting with
   a numeric value.
   'FPM-' for FoLiA-pm. 'FP-' for FoLiA-page, 'FH-' for FoLiA-hocr
   Can bet set witth --prefix
 * FoLiA-stats.cxx:
   - added --collect to usage() and 'man' page
 * FoLiA-correct:
   - added --inputclass and --outputclass parameters (must be different)
   - Don't crash on empty text.

foliautils 0.9.1  17-05-2018
[Ko vd Sloot]
Bug fix release: the tests directory wasn't included in the distribution

foliautils 0.9  16-05-2018
[Ko vd Sloot]
* FoLiA-stats.cxx:
  - added a --collect option, to create files with all n-grams
    together
  - clearer message in FoLiA-stats when no results were found
  - extract text from deeper nodes, if needed
  - fixed out-of-bounds problem
  - now fails when every inputfile fails
* FoLiA-txt:
  - now fails when every inputfile fails
* avoid xml:id's starting with a number. Add "id-" in front.
* added more tests

[Maarten van Gompel]
* added codemeta.json


foliautils 0.8  19-02-2018
[Ko vd Sloot]
* added -R option to FoLiA-collect
* FoLiA-collect now can work in parallel (-t option)
* modernized configuration, whit better Max OSX support (including OpenMP)
* all modules end with an exit code now.
* added more tests to 'make check'
* added output of Type-Token Ratio's (also in degrees)
* several bugfixes.
* code cleanup and refactoring, some speedup too

foliautils 0.7  24-10-2017
[ko vd Sloot]
 * updated and expanded tests
 * fixed offset calculations in FoLiA-hocr, FoLiA-page.cxx
   and FoLiA-alto. We use unicode points now. (needed for folia v1.5 and above)
 * Changed 'modes' in FoLiA-stats, to be a bit more comprehensible
 * fixed problem with metadatatype when 'foreign-data' is present
 * enhanced FoLiA-clean. Still not done...
 * switched to dynamic OMP scheduling in most programs.
   (which process files with probably big differences in processing time)
 * small bugfixes.
 * general cleanup and refactoring

[Maarten van Gompel]
 * Added and improved FoLiA-wordtranslate.cxx

foliautils 0.6  04-04-2017
This is an intermediate release!!

src/FoLiA-idf.cxx
/*
  Copyright (c) 2014 - 2021
  CLST  - Radboud University
  ILK   - Tilburg University

  This file is part of foliautils

  foliautils is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 3 of the License, or
  (at your option) any later version.

  foliautils is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, see <http://www.gnu.org/licenses/>.

  For questions and suggestions, see:
      https://github.com/LanguageMachines/foliautils/issues
  or send mail to:
      lamasoftware (at ) science.ru.nl
*/

#include <getopt.h>
#include <string>
#include <map>
#include <vector>
#include <iostream>
#include <fstream>

#include "ticcutils/FileUtils.h"
#include "ticcutils/CommandLine.h"
#include "ticcutils/StringOps.h"
#include "ticcutils/Unicode.h"
#include "libfolia/folia.h"

#include "config.h"
#ifdef HAVE_OPENMP
#include "omp.h"
#endif

using namespace	std;
using namespace	icu;
using namespace	folia;

bool verbose = false;

void create_idf_list( const map<string, unsigned int>& wc,
		      const string& filename, unsigned int clip ){
  ofstream os( filename );
  if ( !os ){
    cerr << "failed to create outputfile '" << filename << "'" << endl;
    exit(EXIT_FAILURE);
  }
  for ( const auto& cit : wc  ){
    if ( cit.second > clip ){
      os << cit.first << "\t" << cit.second << endl;
    }
  }
#pragma omp critical
  {
    cout << "created IDF list '" << filename << "'" << endl;
  }
}

size_t words_inventory( const Document *doc,
			bool lowercase,
			const string& classname,
			map<string,unsigned int>& wc ){
  vector<Word*> words = doc->words();
  set<string> ws;
  for ( auto const& folia_word : words ){
    string word;
    try {
      if ( lowercase ){
	UnicodeString uword = TiCC::UnicodeFromUTF8( folia_word->str(classname) );
	uword.toLower();
	word = TiCC::UnicodeToUTF8( uword );
      }
      else
	word = folia_word->str(classname);
    }
    catch(...){
#pragma omp critical
      {
	cerr << "missing text for word " << folia_word->id() << endl;
      }
      break;
    }
    ws.insert( word );
  }

  for ( const auto& s : ws ){
#pragma omp critical
    {
      ++wc[s];
    }
  }
  return ws.size();
}

size_t strings_inventory( const Document *doc,
			  bool lowercase,
			  const string& classname,
			  map<string,unsigned int>& wc ){
  vector<String*> words = doc->doc()->select<String>();
  set<string> ws;
  for ( auto const& folia_word : words ){
    string word;
    try {
      UnicodeString uword = folia_word->text(classname);
      if ( lowercase ){
	uword.toLower();
      }
      word = TiCC::UnicodeToUTF8( uword );
    }
    catch(...){
#pragma omp critical
      {
	cerr << "missing text for word " << folia_word->id() << endl;
      }
      break;
    }
    ws.insert( word );
  }

  for ( const auto& s : ws ){
#pragma omp critical
    {
      ++wc[s];
    }
  }
  return ws.size();
}

void usage(){
  cerr << "Usage: [options] file/dir" << endl;
  cerr << "\t FoLiA-idf will produce IDF statistics for a directory of FoLiA files " << endl;
  cerr << "\t--clip\t\t clipping factor. " << endl;
  cerr << "\t\t\t (entries with frequency <= this factor will be ignored). " << endl;
  cerr << "\t--lower\t\t lowercase all words" << endl;
  cerr << "\t--class='name'\t use 'name' as the folia class for <t> nodes. (default is 'current')" << endl;
  cerr << "\t--strings\t search for String nodes (default is Word)" << endl;
  cerr << "\t-t <threads>\n\t--threads <threads> Number of threads to run on." << endl;
  cerr << "\t\t\t If 'threads' has the value \"max\", the number of threads is set to a" << endl;
  cerr << "\t\t\t reasonable value. (OMP_NUM_TREADS - 2)" << endl;
  cerr << "\t-h or --help\t this message " << endl;
  cerr << "\t-V or --version\t show version " << endl;
  cerr << "\t-e\t\t expr: specify the expression all files should match with." << endl;
  cerr << "\t-O\t\t output prefix" << endl;
  cerr << "\t-R\t\t search the dirs recursively (when appropriate)." << endl;
  exit(EXIT_SUCCESS);
}

int main( int argc, char *argv[] ){
  TiCC::CL_Options opts( "vVt:O:Rhe:",
			 "class:,clip:,lower,help,strings,version,threads:" );
  try {
    opts.init( argc, argv );
  }
  catch( TiCC::OptionError& e ){
    cerr << e.what() << endl;
    usage();
    exit( EXIT_FAILURE );
  }
  int clip = 0;
  bool recursiveDirs = false;
  bool lowercase = false;
  string expression;
  string outPrefix;
  string value;
  string classname;
  if ( opts.empty() ){
    usage();
    exit(EXIT_FAILURE);
  }
  if ( opts.extract('h') || opts.extract( "help" ) ){
    usage();
    exit(EXIT_SUCCESS);
  }
  if ( opts.extract('V') || opts.extract( "version" ) ){
    cerr << PACKAGE_STRING << endl;
    exit(EXIT_SUCCESS);
  }
  verbose = opts.extract( 'v' );
  bool do_strings = opts.extract( "strings" );
  opts.extract( 'e', expression );
  recursiveDirs = opts.extract( 'R' );
  opts.extract( 'O', outPrefix );
  if ( opts.extract( 't', value )
       || opts.extract( "threads", value ) ){
#ifdef HAVE_OPENMP
    int numThreads = 1;
    if ( TiCC::lowercase(value) == "max" ){
      numThreads = omp_get_max_threads() - 2;
    }
    else if ( !TiCC::stringTo(value,numThreads) ) {
      cerr << "illegal value for -t (" << value << ")" << endl;
      exit( EXIT_FAILURE );
    }
    omp_set_num_threads( numThreads );
#else
    cerr << "-t option does not work, no OpenMP support in your compiler?" << endl;
    exit( EXIT_FAILURE );
#endif
  }
  if ( opts.extract( "clip", value ) ){
    if ( !TiCC::stringTo( value, clip ) ){
      cerr << "illegal --clip value (" << value << ")" << endl;
      exit( EXIT_FAILURE );
    }
  }
  lowercase = opts.extract( "lower" );
  opts.extract( "class", classname );

  vector<string> fileNames = opts.getMassOpts();
  if ( fileNames.size() == 0 ){
    cerr << "missing input file or directory" << endl;
    exit( EXIT_FAILURE );
  }
  else if ( fileNames.size() > 1 ){
    cerr << "currently only 1 file or directory is supported" << endl;
    exit( EXIT_FAILURE );
  }

  string dirName = fileNames[0];
  fileNames = TiCC::searchFilesMatch( dirName, expression, recursiveDirs );
  size_t toDo = fileNames.size();
  if ( toDo == 0 ){
    cerr << "no matching files found" << endl;
    exit(EXIT_SUCCESS);
  }
  if ( outPrefix.empty() ){
    string::size_type pos = fileNames[0].find( "/" );
    if ( pos != string::npos )
      outPrefix = fileNames[0].substr(0,pos);
    else
      outPrefix = "current";
  }

  if ( toDo > 1 ){
    cout << "start processing of " << toDo << " files " << endl;
  }


  map<string,unsigned int> wc;
  unsigned int wordTotal =0;

#pragma omp parallel for shared(fileNames,wordTotal,wc ) schedule(dynamic)
  for ( size_t fn=0; fn < fileNames.size(); ++fn ){
    string docName = fileNames[fn];
    Document *d = 0;
    try {
      d = new Document( "file='"+ docName + "'" );
    }
    catch ( exception& e ){
#pragma omp critical
      {
	cerr << "failed to load document '" << docName << "'" << endl;
	cerr << "reason: " << e.what() << endl;
      }
      continue;
    }
    size_t count;
    if ( do_strings ){
      count = strings_inventory( d, lowercase, classname, wc );
    }
    else {
      count = words_inventory( d, lowercase, classname, wc );
    }
    wordTotal += count;
#pragma omp critical
    {
      cout << "Processed :" << docName << " with " << count << " unique words."
	   << " still " << --toDo << " files to go." << endl;
    }
    delete d;
  }

  if ( toDo > 1 ){
    cout << "done processsing directory '" << dirName << "' in total "
	 << wordTotal << " unique words were found." << endl;
  }
  cout << "start calculating the results" << endl;
  string filename = outPrefix + ".idf.tsv";
  create_idf_list( wc, filename, clip );
  cout << "done: " << endl;
  return EXIT_SUCCESS;
}
